### PR TITLE
Feature/apmw 120 permission with ccc

### LINKED ISSUE #120: Implement Permission System
Users should have different kinds of permissions (e.g. Not everyone should be allowed to create new persons/locations).
The system should allow users to be part of permission groups (e.g. admin/mod) and by that gain permission to edit buildings, rooms, ... etc.

### Acceptance

- [x] Users can be given permissions
- [ ] Documentation in Github Wiki (how to use)
- [ ] add permissions to rooms and openingtimes (and people?) (https://github.com/hpi-swt2/compass-portal-blue/pull/162#issuecomment-1020297865)

### Done in Sprint 03:

- [x] It can be checked whether a user has permission or not

### DoD

- [x] Tests
- [ ] Documentation
- [x] CI Build passing
- [ ] Peer Review
- [ ] Approval from PO

### PR DESCRIPTION
**Pull request by joh-dah**
closes #120 

This PR implements a permission system.
An admin can manage everything.
A user can create everything and manage (edit/delete) only his owned resources.
When creating a resource, a user gets automatically assigned as owner.
A guest can only read everything.

Assigning the admin role is currently only possible via the console.

image
image

### CODE REVIEW

File: app/models/user.rb
Comment (chrisma) [inline app/models/user.rb:26]: You might want to add a comment on what this line/method accomplishes

File: spec/features/users/user_authorization_spec.rb
Comment (bewee) [inline spec/features/users/user_authorization_spec.rb:5]: Does this test really check people management?

File: spec/features/users/user_authorization_spec.rb
Comment (bewee) [inline spec/features/users/user_authorization_spec.rb:13]: Does this test really check people management?

File: spec/features/users/user_authorization_spec.rb
Comment (bewee) [inline spec/features/users/user_authorization_spec.rb:23]: Does this test really check people management?

File: spec/requests/buildings_spec.rb
Comment (bewee) [inline spec/requests/buildings_spec.rb:27]: Why have you added this here, but not in `locations_spec.rb`?

File: spec/views/buildings/index.html.erb_spec.rb
Comment (bewee) [inline spec/views/buildings/index.html.erb_spec.rb:7]: Ouch, these brackets look quite ugly to me, do we really need them?

**Review by bewee — APPROVED** [review]
Overall, this looks pretty good to me! We need to rebase this branch to get rid of the conflicts. However, I expect #161 to introduce even more conflicts, so we shoud probably wait for this to be merged.
Additionally, we should fix the remaining tests and codefactor issues (Have you tried executing `rubocop -a` locally?)

File: app/models/building.rb
Comment (bewee) [reply app/models/building.rb:2]: Just to clarify: I'm completely fine moving this to a separate feature / PR, just asking :)

File: spec/features/users/user_authorization_spec.rb
Comment (bewee) [reply spec/features/users/user_authorization_spec.rb:5]: Does any test really check people management?

**Comment by bewee** [comment]
> We need to rebase this branch to get rid of the conflicts. However, I expect #161 to introduce even more conflicts, so we shoud probably wait for this to be merged. 

And by the time I submitted the review, it was even already merged :)

**Comment by bewee** [comment]
The problem in `app/models/user.rb#L4` wouldn't occur if we used `has_and_belongs_to_many` as (kind of) suggested above. What do you think about this @Tratori @YoungBinKo @piabeeck ?

File: Gemfile
Comment (chrisma) [inline Gemfile]: Would you want to add a comment to the `Gemfile` on what this particular dependency accomplishes? Maybe a link to how to interact with it would also help others, or a reference to a project file that shows correct usage?

**Comment by bewee** [comment]
What I found out about the failing tests: `spec/requests/rooms_spec.rb` and `spec/requests/openingtimes_spec.rb` are the two problematic files. Interstingly enough, these are also the only two request tests that weren't touched in this PR. At the same time, these are (in contrast to buildings, locations and people) not included in the permission system.
I have no clue what exactly happens here, and especially not why this seamingly completely unrelated error occurs. However, one very simple fix could possibly be to simply add permissions for rooms and openingtimes as well. What do you think? @YoungBinKo @piabeeck @Tratori

**Comment by Tratori** [comment]
> What I found out about the failing tests: `spec/requests/rooms_spec.rb` and `spec/requests/openingtimes_spec.rb` are the two problematic files. Interstingly enough, these are also the only two request tests that weren't touched in this PR. At the same time, these are (in contrast to buildings, locations and people) not included in the permission system. I have no clue what exactly happens here, and especially not why this seamingly completely unrelated error occurs. However, one very simple fix could possibly be to simply add permissions for rooms and openingtimes as well. What do you think? @YoungBinKo @piabeeck @Tratori

I think this would be worth a try, will ad this to the ticket

**Comment by bewee** [comment]
Bug: You can no longer edit your own linked person
Reason: You implemented admin-only people editing permission here: db84f2f928e13a43e3048148bfc2359c48f17d51
Possible solution: Add a habtm relationship between people and users (just as we did between buildings/locations and users), and initialize the auto-generated person with the current user in `app/models/user.rb:20`

**Comment by YoungBinKo** [comment]
> Bug: You can no longer edit your own linked person Reason: You implemented admin-only people editing permission here: [db84f2f](https://github.com/hpi-swt2/compass-portal-blue/commit/db84f2f928e13a43e3048148bfc2359c48f17d51) Possible solution: Add a habtm relationship between people and users (just as we did between buildings/locations and users), and initialize the auto-generated person with the current user in `app/models/user.rb:20`
> 
> What do you think about this @piabeeck @YoungBinKo @Tratori

This [5ff7133](https://github.com/hpi-swt2/compass-portal-blue/pull/162/commits/5ff7133966e04aebfb13a4067f9f5da634e2f30d) looks pretty good! Should we add now then tests for people? @bewee

File: app/controllers/people_controller.rb
Comment (taminob) [inline app/controllers/people_controller.rb:1]: Why is here in PeopleController no `@person.owners = [current_user]`?

File: app/models/user.rb
Comment (taminob) [inline app/models/user.rb:36]: not yet necessary because there aren't multiple roles right now, right?
Probably copies from here: https://github.com/CanCanCommunity/cancancan/blob/develop/docs/role_based_authorization.md

File: app/views/buildings/index.html.erb
Comment (taminob) [inline app/views/buildings/index.html.erb:27]: What about an `if can? :create, Person` for the "New Person" button? Anyone an idea, how this would work? Tried this and didn't seem to work

File: app/views/buildings/new.html.erb
Comment (taminob) [inline app/views/buildings/new.html.erb:5]: Why is this here?

File: app/views/locations/index.html.erb
Comment (taminob) [inline app/views/locations/index.html.erb:29]: Same here with the "New Location" button

File: db/migrate/20220125182817_remove_user_id_from_buildings.rb
Comment (taminob) [inline db/migrate/20220125182817_remove_user_id_from_buildings.rb:5]: Good intention, but I think we should maybe just remove these migrations and the ones which add the columns? In my opinion, as long as the migration didn't enter dev, it doesn't matter if you have to drop the database

File: spec/features/locations/edit_spec.rb
Comment (taminob) [inline spec/features/locations/edit_spec.rb]: I agree with Benedikt, we should remove this line since it isn't used

**Review by taminob — COMMENTED** [review]
I pushed some changes and included rooms in the permission system. Someone should have a look at the changes though if I included everything and didn't miss anything.

File: spec/features/locations/edit_spec.rb
Comment (taminob) [reply spec/features/locations/edit_spec.rb]: removed this already

File: app/controllers/people_controller.rb
Comment (taminob) [reply app/controllers/people_controller.rb:1]: done

File: app/views/buildings/index.html.erb
Comment (taminob) [reply app/views/buildings/index.html.erb:27]: done

File: app/views/locations/index.html.erb
Comment (taminob) [reply app/views/locations/index.html.erb:29]: done

File: app/models/user.rb
Comment (chrisma) [reply app/models/user.rb:36]: Any chance of adding some explanatory comments for this?
`((roles_mask.to_i || 0) & (2**ROLES.index(r))).zero?` might not be immediately obvious

Comment (bewee) [reply app/models/user.rb:36]: We are too waiting for the PR owners :)

Comment (YoungBinKo) [reply app/models/user.rb:36]: > not yet necessary because there aren't multiple roles right now, right? Probably copied from here: https://github.com/CanCanCommunity/cancancan/blob/develop/docs/role_based_authorization.md#many-roles-per-user

Yes, it's from there and it's not completed. I think we can for now delete this, unless PO has some other Roles to apply this.

**Comment by taminob** [comment]
Currently, the permission system allows everyone to create/edit/show any openingtime.
We propose to leave this as it is and remove the routes for this (openingtimes, openingtimes/new, etc.) and thus make this impossible for the user.
Introducing a permission system for openingtimes would be too complicated to add in this PR since some kind of synchronization system would be required to sync owners with the corresponding person/room.

File: app/models/user.rb
Comment (taminob) [reply app/models/user.rb]: not needed anymore, deleted

File: app/models/user.rb
Comment (taminob) [reply app/models/user.rb]: We decided to remove this for now. Should be added in a separate PR when multiple roles are required.

**Comment by taminob** [comment]
@Tratori as soon as you approve, this is ready to be merged

**Review by bewee — APPROVED** [review]

**Review by Tratori — APPROVED** [review]
I'm happy we finished this beast of a task :D
good job everyone!

**Pull request merged**
